### ISSUE TITLE
Increasing data augmentation throughput/performance

### ISSUE THREAD

**Issue by ORippler** [MIC-DKFZ/batchgenerators#5]
Dear DKFZ,

depending on the batch-size and augmentation types, I experience performance problems, where on-the-fly augmentation is significantly slowing down my training procedure. I.e. CPU is working at 100% load, but still cannot provide/augment samples fast enough for the GPU to crunch through them.

I did some profiling of your code and noticed that this was especially the case when `scipy.ndimage.interpolation.map_coordinates` is involved. I already change the interpolation order to 1, but still experience significant slowdown for high batch-sizes/large patch-volumes. Do you have any suggestions for improvements?

I thought about maybe integrating `ITK`, which now has a [python wrapper](https://itkpythonpackage.readthedocs.io/en/latest/Quick_start_guide.html#mixing-itk-and-numpy), that allows memory sharing between `np.ndarray` and `itk.image` objects, and performing the actual deformation in itk with [`itk.WarpImageFilter`](https://itk.org/ITKExamples/src/Filtering/ImageGrid/WarpAnImageUsingADeformationField/Documentation.html?highlight=deformation%20field). What are your thoughts on this? Did you maybe try it out already?

**Comment by FabianIsensee** [comment]
Hi,
I have not tried this, but I hardly doubt this will improve performance. map_coordinates uses a c++ backend as well. It's going as fast as it gets.
What augmentation are you using that makes use of map_coordinates? I think that must be SpatialTransform right? Could you please share the dimensionality of your tensors (both 'data' and 'seg' if you have a segmentation problem)?
I have to leave work now, I will post some performance suggestions tomorrow.
Best,
Fabian

**Comment by Schnabbel** [comment]
Hi,

yes, I am indeed using `SpatialTransform` (rotation, scaling, elastic deformation, random crop) in both 2D and 3D cases.
For 3D, my patch-size for 'data'  has dimensionality of (1,1,256,256,128) in (b,c,x,y,z) and 'seg' a shape of (1,3,256,256,128).
For 2D, my patch-size for 'data' has shape (32,1,224,224) and 'seg' is not present, as it is a classification problem I am trying to solve.
Thanks in advance for your help.

Best,

**Comment by FabianIsensee** [comment]
Alright let's see.
As a side note: The color channel in 'seg' are in fact one hot encodings of your classes? You don't need one hot encodings, it will do the augmentation properly now and you will not get interpolation artifacts (it will now do a one hot encoding internally if needed).

Your patch size (segmentation) is very large. My guess is you are using a U-Net or FCN like architecture. In order to process patches this large, your architecture probably needs to be quite thin (talking 16 base feature maps in a UNet) (I am assuming you have 12 GB of GPU memory). A thin network would of course put more strain on the CPU side of things because it can process batches faster.
It is not unexpected that map_coordinates will take a while augmenting patch sizes that large. From my experience with BraTS (2,4,128,128,128) tensors the CPU (they have 2x as many voxels as yours) this does not have to be a problem though. Here are a few things you can do:

1) pull the most recent version. I added some changes to interpolate_img last week
2) don't use a one hot encoding of your segmentation (if possible) and set order_seg=0 in SpatialTransform. Otherwise the time needed to interpolate your segmentation will increase linearly with the number of labels (I have never been able to demonstrate that it is actually better to do other types of interpolation on the labels)
3) Make sure you are using the multithreaded augmenter correctly. Use MultiThreadedAugmenter. It is faster than AlternativeMultiThreadedAugmenter. I have 8C/16T and am using n_jobs=8 or so. num_cached_per_queue does not need to be more than 1, maybe 2.
4) You probably don't want to always augment batches with SpatialTransform. It is actually beneficial to also show unmodified examples. Use RndTransform for that. Example:
`
RndTransform(my_spatial_trnasform, prob=0.5, alternative_transform=RandomCropTransform(my_patch_size))
`
That will augment only 50% of your batches.

This should help for both your 2D and 3D network. I am a bit surprised though that you have problems with 2D. I am using batches of size (89,1,192,160) for another segmentation problem of mine and the CPU is not doing much.

Let me know if that helped =)
Best,
Fabian

**Comment by ORippler** [comment]
Hey,

yes, the classes for the 3D segmentation problem are in 'one_hot' format, and I identified the bottleneck to be disk I/O for this task, as i do asynchrous loading from disk during training :)

Regarding the 2D classification, I load datasets into memory prio to training, as they are small enough. Here, I:
* Did pull the most recent version, though i am still using `DataLoaderBase` class (will switch to `SlimDataLoaderBase` when i find the time.)
* Probably am not using `MultiThreadedAugmenter` correctly. The reason for this is I only get a speedup of around 0.2 when setting num_processes to 8 compared to 1 (e.g. 2.6 batches/s for 8 processes compared to 2.1 for 1 process). Maybe you could give some expected values for the speed-up factor?

If I can fix this, i guess the computational cost of data augmentations with `Spatial_Transform` will be no longer an issue.

Also, maybe a little bit off-topic, but in the `multithreaded_dataloding` example `DummyDLWithShuffle` does not shuffle its data (e.g. multiple iterations over mt yield identical batches).

Thanks again for your help, it's much appreciated!

**Comment by FabianIsensee** [comment]
Hi,
since you said that your CPU was at 100% I did not mention HDD bottleneck as one of the causes. That is of course often a problem - I solve that by storing everything on SSD.

Can you please post your code where MultiThreadedAugmenter does not give a good speedup? For me it will always almost give perfect scaling. There are some things that will slow it down (such as large tensors being transferred between threads) but it should never be as bad as you are describing. I can only figure out what's going on once I see your data augmentation code.

Thanks for the hint with DummyDLWithShuffle. I will look into this.

Best,
Fabian

**Comment by FabianIsensee** [comment]
I just tried to reproduce your problem with DummyDLWithShuffle. It works for me. Here are the three lists that are printed if I run the script. As you can see the 'data' is returned in a different order each time I iterate over it:

`[26, 86, 2, 55, 75, 93, 16, 73, 54, 95, 53, 92, 78, 13, 7, 30, 22, 24, 33, 8, 43, 62, 3, 71, 45, 48, 6, 99, 82, 76, 60, 80, 90, 68, 51, 27, 18, 56, 63, 74, 1, 61, 42, 41, 4, 15, 17, 40, 38, 5, 91, 59, 0, 34, 28, 50, 11, 35, 23, 52, 10, 31, 66, 57, 79, 85, 32, 84, 14, 89, 19, 29, 49, 97, 98, 69, 20, 94, 72, 77, 25, 37, 81, 46, 39, 65, 58, 12, 88, 70, 87, 36, 21, 83, 9, 96, 67, 64, 47, 44]

[25, 39, 68, 37, 83, 24, 18, 81, 89, 85, 21, 74, 11, 0, 59, 90, 80, 4, 72, 53, 2, 97, 64, 66, 8, 27, 9, 99, 17, 63, 84, 44, 28, 96, 88, 1, 38, 52, 71, 51, 58, 34, 94, 30, 46, 61, 15, 36, 6, 47, 41, 50, 45, 75, 23, 5, 62, 12, 55, 98, 60, 32, 19, 42, 40, 70, 54, 10, 26, 87, 35, 3, 31, 57, 73, 67, 13, 14, 65, 7, 76, 82, 92, 33, 43, 91, 48, 16, 29, 20, 86, 22, 79, 77, 93, 69, 95, 49, 78, 56]

[33, 84, 62, 8, 80, 66, 68, 99, 17, 0, 56, 79, 13, 59, 25, 97, 37, 63, 32, 7, 1, 74, 43, 46, 57, 24, 27, 14, 73, 19, 70, 39, 72, 6, 38, 65, 18, 20, 22, 21, 11, 75, 16, 23, 69, 28, 53, 9, 60, 5, 85, 95, 4, 98, 12, 34, 40, 61, 49, 89, 3, 93, 86, 78, 48, 76, 41, 45, 54, 29, 35, 15, 26, 87, 82, 55, 96, 71, 50, 94, 83, 10, 51, 52, 2, 44, 42, 36, 91, 77, 47, 88, 81, 67, 92, 30, 64, 31, 90, 58]
`

**Comment by ORippler** [comment]
Hi,

i rebuild my Conda environment and could not reproduce the DummyDLWithShuffle behavior, so I guess that one is on me, sorry.

I setup a MWE for the 2D dataloader case, showcasing the problem of improper speedup. I reproduced the behavior in a docker environment, and have attached the dockerfile as well as the MWE and .pstats generated by cProfile in a [zip](https://github.com/MIC-DKFZ/batchgenerators/files/2271043/MWE.zip).

Looking forward to your input =)

Best,

**Comment by FabianIsensee** [comment]
Hi,
I simplified your script to make it easier to find out what's going on (basically i scrapped your data loader and replaced it with DummyLoader - it produces the same size of array than yours did).

```
from batchgenerators.dataloading import SingleThreadedAugmenter, MultiThreadedAugmenter
from batchgenerators.transforms.color_transforms import \
    BrightnessMultiplicativeTransform
from batchgenerators.transforms.spatial_transforms import SpatialTransform
from batchgenerators.transforms.abstract_transforms import Compose
from batchgenerators.transforms.sample_normalization_transforms import \
    MeanStdNormalizationTransform
from batchgenerators.dataloading.data_loader import SlimDataLoaderBase
import numpy as np
from time import time


class DummyLoader(SlimDataLoaderBase):
    def __init__(self):
        super(DummyLoader, self).__init__(None, None, None)

    def generate_train_batch(self):
        return {'data': np.random.random((32, 3, 224, 224))}


transforms = Compose([
    BrightnessMultiplicativeTransform(multiplier_range=(0.7, 1.3),
                                      per_channel=True),

    SpatialTransform(patch_size=(224, 224),
                     do_elastic_deform=True,
                     alpha=(90., 750.),
                     sigma=(9., 11.),
                     do_scale=True,
                     random_crop=False,
                     do_rotation=False,
                     order_data=1,
                     border_mode_data='reflect'),

    MeanStdNormalizationTransform(mean=[0.485, 0.456, 0.406],
                                  std=[0.229, 0.224, 0.225])
])


single_threaded_gen = SingleThreadedAugmenter(DummyLoader(), transforms)
multi_threaded_gen_one_thread = MultiThreadedAugmenter(DummyLoader(), transforms, 1, 1, None)
multi_threaded_gen_eight_threads = MultiThreadedAugmenter(DummyLoader(), transforms, 8, 1, None)

num_batches_warmup = 10
num_batches_run = 50


####### SingleThreadedAugmenter #######
# warumup
_ = [next(single_threaded_gen) for _ in range(num_batches_warmup)]
# run
start = time()
_ = [next(single_threaded_gen) for _ in range(num_batches_run)]
end = time()
print("Generated %d batches with SingleThreadedAugmenter in %f seconds; %f s/batch" % (num_batches_run, end - start, (end - start) / num_batches_run))


####### MultiThreadedAugmenter (1 thread) #######
# warumup
_ = [next(multi_threaded_gen_one_thread) for _ in range(num_batches_warmup)]
# run
start = time()
_ = [next(multi_threaded_gen_one_thread) for _ in range(num_batches_run)]
end = time()
print("Generated %d batches with MultiThreadedAugmenter (1 thread) in %f seconds; %f s/batch" % (num_batches_run, end - start, (end - start) / num_batches_run))


####### MultiThreadedAugmenter (8 threads) #######
# warumup
_ = [next(multi_threaded_gen_eight_threads) for _ in range(num_batches_warmup)]
# run
start = time()
_ = [next(multi_threaded_gen_eight_threads) for _ in range(num_batches_run)]
end = time()
print("Generated %d batches with MultiThreadedAugmenter (8 threads) in %f seconds; %f s/batch" % (num_batches_run, end - start, (end - start) / num_batches_run))
```
This script will give the following output on my machine:

> Generated 50 batches with SingleThreadedAugmenter in 33.483667 seconds; 0.669673 s/batch
> Generated 50 batches with MultiThreadedAugmenter (1 thread) in 35.732181 seconds; 0.714644 s/batch
> Generated 50 batches with MultiThreadedAugmenter (8 threads) in 5.994122 seconds; 0.119882 s/batch
> 

The loss in speed with MultiThreadedAugmenter (threads=1) is expected because we now have thread communication overhead. MultiThreadedAugmenter (threads=8)  however has a ~7x speedup relative to MultiThreadedAugmenter (threads=1), which I think is very acceptable considering that I have other things running on my computer as well.

As you can see I could not (yet) reproduce your problem. I think you must have a problem in your data loader (maybe I/O?). As I don't know how you are loading things exactly and what else your data loader does I cannot help you. If you can compile a minimalistic example that will reproduce your problem I may be able to help though.

Best,
Fabian

**Comment by ORippler** [comment]
Okay, so I finally found the solution to the problem.
It was that we used packages provided by `conda` in our dockerfile/installation (e.g. `conda install numpy` etc.).

As i could not narrow it down to a single package (multiple ones lead to the slowdown), I switched to installing only using `pip`, and can now get the expected performance gains. I did not narrow the cause down further, but would suggest you to advise people not to use `conda` packages together with `MultiThreadedAugmenter`, but instead install them using `pip` =)

[I attached the old, broken dockerfile as well as the new, working one](https://github.com/MIC-DKFZ/batchgenerators/files/2278327/dockerfiles.zip). If i find the time, I will open a git-issue in the conda repository, so they can have a look at it.

Have a nice weekend, thanks for your help and again: Thanks for providing such a nice framework to the community!
